src/merkle/path.rs

```diff
@@ -1,4 +1,88 @@
-use super::{BTreeMap, MerkleError, Rpo256, Vec, Word, ZERO};
+use super::{vec, BTreeMap, MerkleError, Rpo256, Vec, Word, ZERO};
+use core::ops::{Deref, DerefMut};
+
+// MERKLE PATH
+// ================================================================================================
+
+/// A merkle path container, composed of a sequence of nodes of a Merkle tree.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct MerklePath {
+    nodes: Vec<Word>,
+}
+
+impl MerklePath {
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Creates a new Merkle path from a list of nodes.
+    pub fn new(nodes: Vec<Word>) -> Self {
+        Self { nodes }
+    }
+
+    // PROVIDERS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns the depth in which this Merkle path proof is valid.
+    pub fn depth(&self) -> u8 {
+        self.nodes.len() as u8
+    }
+
+    /// Verify the Merkle opening proof towards the provided root.
+    ///
+    /// It will assert `input` exists on a Merkle tree, indexed by `index` for an arbitrary depth.
+    pub fn verify(&self, mut index: u64, node: Word, root: &Word) -> bool {
+        let computed = self.nodes.iter().copied().fold(node, |node, sibling| {
+            // build the input node, considering the parity of the current index.
+            let is_right_sibling = (index & 1) == 1;
+            let input = if is_right_sibling {
+                [sibling.into(), node.into()]
+            } else {
+                [node.into(), sibling.into()]
+            };
+            // compute the node and move to the next iteration.
+            index >>= 1;
+            Rpo256::merge(&input).into()
+        });
+        root == &computed
+    }
```

Comment:
I wonder if we should split this into two functions: one to compute the root and the other to verify the path. The compute root function could have the following signature:
```Rust
pub fn compute_root(&self, mut index: u64; node: Word) -> Word
```
And then, `verify()` function would look like this:
```Rust
pub fn verify(&self, index: u64, node: Word, root: &Word) -> bool {
    let computed_root = self.compute_root(index, node);
    root == computed_root
}
```

Lastly, I think the comment here could be updated similarly to what I suggested in the other PR.

Reply:
I made it a bit smaller to reuse the arguments of the function as arguments of the sentence.

```rust
/// Verifies the Merkle opening proof towards the provided root.
///
/// Returns `true` if `node` exists at `index` in a Merkle tree with `root`.

```

wdyt?

Reply:
Looks good!